llvm/tools/repo-fragments/RepoFragments.cpp

```diff
@@ -58,17 +58,59 @@ raw_ostream &operator<<(raw_ostream &OS, NameAndDigest const &ND) {
 // Returns an object that can be written to an ostream to produce the requested
 // output.
 static NameAndDigest makeNameAndDigest(bool DigestOnly, const StringRef &Name,
-                                       pstore::index::digest const &Digest) {
+                                       const pstore::index::digest &Digest) {
   if (DigestOnly) {
     return NameAndDigest{Digest};
   }
   return NameAndDigest{Name, Digest};
 }
 
-// Clamps the size_t input value to the maximum unsigned value.
-static constexpr unsigned clamp_to_unsigned(size_t S) {
-  constexpr size_t largest = std::numeric_limits<unsigned>::max();
-  return static_cast<uint16_t>(S > largest ? largest : S);
+/// Dumps selected entries from a compilation.
+///
+/// \tparam NameIterator  An input iterator whose value-type is compatible with
+///   StringRef.
+/// \param Db  The owning database instance.
+/// \param Compilation  The compilation whose members are to be dumped.
+/// \param First  The start of the range of objects to be dumped.
+/// \param Last  The end of the range of objects to be dumped.
+/// \param Dump  A function which is called to dump an individual name. It must
+///   have a signature compatible with std::function<void(const
+///   pstore::repo::definition &)>.
+/// \returns False if one of the names in the [First, Last) range was not found
+///   in the compilation. True otherwise.
+template <typename NameIterator, typename DumpFunction>
+static bool dumpSelected(const pstore::database &Db,
+                         const pstore::repo::compilation &Compilation,
+                         NameIterator First, NameIterator Last,
+                         DumpFunction Dump) {
+  using pstore::indirect_string;
+  using pstore::repo::definition;
+
+  bool Okay = true;
+
+  // The collection of names defined by this compilation.
+  std::unordered_map<indirect_string, const definition *> Names{
```

Comment:
(Before you ask, DenseMap<> isn’t be used here because it doesn’t know how to use a key of type indirect_string. Writing the traits-class to teach it is possible but would require changing the indirect_string class itself to allow for “empty” and ”tombstone” values. That’s beyond the scope of this change.)